SpeechResponder/CustomFunctions/ShipCallsign.cs

```diff
@@ -17,36 +20,87 @@ public class ShipCallsign : ICustomFunction
+
+            string result = "";
 
-            string result;
-            if (ship != null)
+            if (string.IsNullOrEmpty(ship?.manufacturer)) { return result; }
+
+            // First, obtain the phonetic manufacturer. This may be the complete name or only a partial name, depending on the manufacturer.
+            var phoneticmanufacturer = ShipDefinitions.ManufacturerPhoneticNames.FirstOrDefault(m => m.Key == ship.manufacturer).Value;
+
+            switch (ship.manufacturer)
             {
-                if (EDDI.Instance.Cmdr != null && EDDI.Instance.Cmdr.name != null)
-                {
-                    // Obtain the first three characters
-                    string chars = new Regex("[^a-zA-Z0-9]").Replace(EDDI.Instance.Cmdr.name, "").ToUpperInvariant().Substring(0, 3);
-                    result = ship.phoneticmanufacturer + " " + Translations.ICAO(chars);
-                }
-                else
-                {
-                    if (string.IsNullOrEmpty(ship.phoneticmanufacturer))
+                case "Core Dynamics":
+                case "Gutamaya":
+                case "Saud Kruger":
+                case "Zorgon Peterson":
+                    // Full names
+                    foreach (Translation item in phoneticmanufacturer) { result += "<phoneme alphabet=\"ipa\" ph=\"" + item.to + "\">" + item.from + "</phoneme> "; }
+                    break;
+                case "Lakon Spaceways":
+                    // First word names
+                    result += "<phoneme alphabet=\"ipa\" ph=\"" + phoneticmanufacturer.First().to + "\">" + phoneticmanufacturer.First().from + "</phoneme> ";
+                    break;
+                case "Faulcon DeLacy":
+                    // Last word names
+                    result += "<phoneme alphabet=\"ipa\" ph=\"" + phoneticmanufacturer.Last().to + "\">" + phoneticmanufacturer.Last().from + "</phoneme> ";
+                    break;
+                default:
+                    // Ship model isn't here
+                    if (ship.phoneticmanufacturer != null && ship.phoneticmanufacturer.Any())
                     {
-                        result = "unidentified ship";
+                        foreach (Translation item in phoneticmanufacturer) { result += "<phoneme alphabet=\"ipa\" ph=\"" + item.to + "\">" + item.from + "</phoneme> "; }
                     }
                     else
                     {
-                        result = "unidentified " + ship.phoneticmanufacturer + " " + ship.phoneticmodel;
+                        result += ship.manufacturer;
                     }
-                }
+                    break;
             }
-            else
+
+            string Get3LeadingCharacters(string input)
             {
-                result = "unidentified ship";
+                // Obtain the first three characters of the input string, zero padded and converted to ICAO (e.g. "A" becomes "Alpha Zero Zero")
```

Comment:
Test coverage would be good here too. I can think of plenty of edge cases like "A-1-B" or "--A--".

Reply:
Done.